File: ios/Classes/UserIdProvider.swift
import Metrix
import Flutter

class UserIdStreamHandler: NSObject, FlutterStreamHandler {
    private var userIdEventSink: FlutterEventSink? = nil

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        userIdEventSink = events
        MetrixClient.setUserIdListener { (userId: String) in
            self.userIdEventSink?(userId)
        }
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        userIdEventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/ir/metrix/flutter/UserIdProvider.kt
package ir.metrix.flutter

import io.flutter.plugin.common.EventChannel
import ir.metrix.attribution.UserIdListener

class UserIdReader: EventChannel.StreamHandler {
    private var userIdEventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        userIdEventSink = events
        ir.metrix.attribution.MetrixAttribution.setUserIdListener(
            object : UserIdListener {
                override fun onUserIdReceived(userId: String) {
                    userIdEventSink?.success(userId)
                }
            })
    }

    override fun onCancel(arguments: Any?) {
        userIdEventSink = null
    }
}